aten/src/ATen/templates/TensorBody.h

```diff
@@ -552,6 +552,26 @@ class CAFFE2_API Tensor {
     return impl_->grad();
   }
 
+  /// This function returns the forward gradient for this Tensor.
+  /// You can set this value to start computing forward gradients using
+  /// set_fw_grad below.
+  const Tensor& fw_grad() const {
+    return impl_->fw_grad();
+  }
+
+  /// This function can be used to set the value of the forward grad.
+  /// Note that the given value might not be used directly if this Tensor already
+  /// has a forward grad or if it is a view of another Tensor.
+  void set_fw_grad(Tensor& new_grad) {
```

Comment:
Should we really be encouraging users to mutably set the forward gradient here? It seems better to provide a functional interface by default for setting these values.

Reply:
Do you mean the argument should be a `const Tensor&`? Yes definitely a oversight while writing the function!

Reply:
Not just that; if I want to construct a dual number, the default way of doing so should be via some *constructor* that (non-mutably) constructs the dual number, rather than by taking an existing tensor and then mutating it so that it has a `fw_grad`.

I realize you have discussed this in the design in this sentence:

> In the first iteration of this feature with no “user friendly” API, when you want to compute J v, you need to set t.fw_grad = v, then perform your computations. You can then read on the output Tensor out.fw_grad that will contain the result of this computation.

But I think this is a step too far, especially given that modern idiomatic PyTorch usage of `requires_grad` involves setting it in the constructor, rather than as a mutable operation after the fact.